File: src/main/scala/iant/zio/demo/PhoneBook.scala
package iant.zio.demo

import zio._
import zio.console._

object PhoneBook extends App {

  def run(args: List[String]): ZIO[ZEnv, Nothing, Int] =
    program.fold(_ => 1, _ => 0)

  val program: ZIO[Console, Throwable, Unit] = for {
    db <- Ref.make(Map.empty[String, String]) // initialize DB
    _  <- cmdLoop(db)
          .catchAll { err =>
            putStrLn(s"Action failed: ${err.getMessage}".colored(scala.Console.RED))
          }.repeat(Schedule.forever)
  } yield()

  implicit class Colorizer(val s: String) extends AnyVal {
    def colored(color: String = scala.Console.BLUE) =
      s"$color$s${scala.Console.RESET}"
  }

  type PhoneBook = Ref[Map[String, String]]

  private def cmdLoop(db: PhoneBook) = for {
    _   <- putStrLn("""Enter action:
                      |    c(reate) name number
                      |    r(ead) [name]
                      |    u(pdate) name number
                      |    d(delete) name
                    """.stripMargin)
    cmd <- getStrLn
    _   <- runCmd(db, cmd)
  } yield ()

  private def runCmd(db: PhoneBook, cmd: String) = {

    def create(name: String, number: String) = db.modify //Not idempotent!!!
      { records =>
        records.get(name) match {
          case Some(_) => (true, records)
          case None    => (false, records + ((name, number)))
        }
      }.flatMap { exists =>
        if (exists) ZIO.fail(
          new RuntimeException(s"Record for '$name' already exists!")
        ) else putStrLn(s"Record created - $name: $number".colored())
      }
    def read(name: String) = db.get.flatMap { records =>
        val msg = records.get(name).fold(s"Record for '$name' does not exist!")(
                  number => s"Record retrieved - $name: $number")
        putStrLn(msg.colored())
      }
    def readAll = db.get.flatMap { records =>
      putStrLn(records.mkString(", ").colored())
    }
    def update(name: String, number: String) =
      db.update{ _ + ((name, number)) } *>
      putStrLn(s"Record created - $name: $number".colored())
    def delete(name: String) =
      db.update(_ - name) *>
      putStrLn(s"Record deleted - $name".colored())

    Task
      .effect(cmd.split("\\s+").toList)
      .collectM(new RuntimeException("Invalid command: " + cmd)) {
        case c :: name :: num :: Nil if "create".startsWith(c) => create(name, num)
        case r :: name        :: Nil if "read".startsWith(r)   => read(name)
        case r                :: Nil if "read".startsWith(r)   => readAll
        case u :: name :: num :: Nil if "update".startsWith(u) => update(name, num)
        case d :: name        :: Nil if "delete".startsWith(d) => delete(name)
      }
    }
}

File: phonebook.py
import os

os.system("")  # Windows hack


class Style:
    RED = '\033[31m'
    BLUE = '\033[34m'
    RESET = '\033[0m'


db = {}  # Initialize database


def __colored(s: str, color: str = Style.BLUE) -> str:
    return f"{color}{s}{Style.RESET}"


def __cmd_loop():
    while True:
        print("Enter action:\n"
              "\tc(reate) name number\n"
              "\tr(ead) [name]\n"
              "\tu(pdate) name number\n"
              "\td(delete) name\n")
        command = input()
        try:
            __run_cmd(command)
        except BaseException as err:
            print(__colored(f"Action failed: {err}", Style.RED))


def __run_cmd(command: str):
    cmd, *tail = command.split(" ")

    if 'create'.startswith(cmd):
        __create(*tail)
    elif 'read'.startswith(cmd):
        if tail:
            __read(tail[0])
        else:
            __read_all()
    elif 'update'.startswith(cmd):
        name, number = tail
        __update(name, number)
    elif 'delete'.startswith(cmd):
        __delete(tail[0])
    else:
        raise ValueError("wrong verb")


def __create(name: str, number: str):
    if name in db:
        raise ValueError(f"Record for '{name}' already exists!")
    else:
        db[name] = number
        print(__colored(f"Record created - {name}: {number}"))


def __read(name: str):
    if name in db:
        number = db[name]
        print(__colored(f"Record retrieved - {name}: {number}"))
    else:
        raise ValueError(f"Record for '{name}' does not exist!")


def __read_all():
    print(__colored(str(db)))


def __update(name: str, number: str):
    db[name] = number
    print(__colored(f"Record updated - {name}: {number}"))


def __delete(name: str):
    db.pop(name, None)
    print(__colored(f"Record '{name}' deleted"))


__cmd_loop()
